src/main/java/com/amazon/opendistroforelasticsearch/security/support/WildcardMatcher.java

```diff
@@ -30,626 +30,503 @@
-                        if (!checkRegionMatches(candidate, textIdx, wcs[wcsIdx])) {
-                            // couldnt match token
-                            break;
-                        }
-                    }
-
-                    // matched text token, move text index to end of matched token
-                    textIdx += wcs[wcsIdx].length();
-                    anyChars = false;
+                    if (wild_backup == -1) return false;
+                    i = ++text_backup;
+                    j = wild_backup;
                 }
-
-                wcsIdx++;
             }
+            while (j < m && pattern.charAt(j) == '*') j++;
+            return j >= m;
+        }
 
-            // full match
-            if (wcsIdx == wcs.length && textIdx == candidate.length()) {
-                return true;
-            }
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            SimpleMatcher that = (SimpleMatcher) o;
+            return pattern.equals(that.pattern);
+        }
 
-        } while (backtrack.size() > 0);
+        @Override
+        public int hashCode() {
+            return pattern.hashCode();
+        }
 
-        return false;
+        @Override
+        public String toString(){ return pattern; }
     }
 
-    /**
-     * Splits a string into a number of tokens.
-     * The text is split by '?' and '*'.
-     * Where multiple '*' occur consecutively they are collapsed into a single '*'.
-     *
-     * @param text  the text to split
-     * @return the array of tokens, never null
-     */
-    private static String[] splitOnTokens(final String text) {
-        // used by wildcardMatch
-        // package level so a unit test may run on this
-
-        if (text.indexOf('?') == NOT_FOUND && text.indexOf('*') == NOT_FOUND) {
-            return new String[] { text };
-        }
-
-        final char[] array = text.toCharArray();
-        final ArrayList<String> list = new ArrayList<>();
-        final StringBuilder buffer = new StringBuilder();
-        char prevChar = 0;
-        for (final char ch : array) {
-            if (ch == '?' || ch == '*') {
-                if (buffer.length() != 0) {
-                    list.add(buffer.toString());
-                    buffer.setLength(0);
-                }
-                if (ch == '?') {
-                    list.add("?");
-                } else if (prevChar != '*') {// ch == '*' here; check if previous char was '*'
-                    list.add("*");
-                }
-            } else {
-                buffer.append(ch);
-            }
-            prevChar = ch;
+    // MatcherCombiner is a combination of a set of matchers
+    // matches if any of the set do
+    // Empty MultiMatcher always returns false
+    private static final class MatcherCombiner extends WildcardMatcher {
```

Comment:
Where such matcher will be used?

Reply:
Have seen `matchAll` mentioned [here](https://github.com/opendistro-for-elasticsearch/security/blob/6d07a03422da1843cba8bf9cdab709d19160ddb3/src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/ConfigModelV6.java#L536) and [here](https://github.com/opendistro-for-elasticsearch/security/blob/00da48ae557c00719cd5ed947d87e40344989c41/src/main/java/com/amazon/opendistroforelasticsearch/security/securityconf/ConfigModelV7.java#L520).

Reply:
It is a different `matchAll`. It is `true` if all candidates matches a pattern or patterns in case of `MatcherCombiner`, compared to a candidate matching all pattern or matching any pattern.

Reply:
Thanks for the clarification.